treematcher.py

```diff
@@ -1,286 +1,178 @@
-    """
-    _syntax_tuples = [
-        ("@", "__target"),
-        ("__target.size", "len(__target.get_leaves())"),
-        ("__target.leaves", "[node.name for node in __target.get_leaves()]"),
-        ("__target.contains_species", "[n.species for n in __target.traverse()]"),
-    ]
-
-    _syntax_tuples_with_cache = [
-        ("@", "__target"),
-        ("__target.size", "len(temp_leaf_cache[__target])"),
-        ("__target.contains_species", "[n.species for n in temp_leaf_cache[__target]]"),
-        ("__target.leaves", "[n.name for n in temp_leaf_cache[__target]]"),
-
-    ]
+class TreePatternCache(object):
+    def __init__(self, tree):
+        """ Creates a cache for attributes that require multiple tree
+        traversal when using complex TreePattern queries.
 
-    def __init__(self, *args, **kargs):
-        kargs["format"] = 1
-
-        is_exact = False
-
-        self.all_node_cache = None
-        self.shortcut_functions = []
-
-
-
-        try:
-            self._extra_functions = kargs.pop('functions')
+        :param tree: a regular ETE tree instance
+         """
+        # Initialize cache (add more stuff as needed)
+        self.leaves_cache = tree.get_cached_content()
+        self.all_node_cache = tree.get_cached_content(leaves_only=False)
 
-        except KeyError:
-            self._extra_functions = None
+    def get_cached_attr(self, attr_name, node, leaves_only=False):
+        """
+        easy access to cached attributes on trees.
 
-        try:
-            self.cache_flag = kargs.pop('cache')
+        :param attr_name: any attribute cached in tree nodes (e.g., species,
+         name, dist, support, etc.)
 
-        except KeyError:
-            self.cache_flag = False
+        :param node: The pattern tree containing the cache
 
+        :leaves_only: If True, return cached values only from leaves
 
-        if len(args) > 0:
-            pattern_string = args[0].strip()
+        :return: cached values for the requested attribute (e.g., Homo sapiens,
+         Human, 1.0, etc.)
 
-            if pattern_string.find('@') == -1:
-                is_exact = True
-            else:
-                #cached_searches = ['.leaves', ".contains_species", ".contains", ".size"]
+        """
+        print "USING CACHE"
+        cache = self.leaves_cache if leaves_only else self.all_node_cache
+        values = [getattr(n, attr_name, None) for n in cache[node]]
+        return values
 
-                shortcut_functions=["contains_species", "contains_duplication", "contains_speciation", "contains"]
+    def get_leaves(self, node):
+        return self.leaves_cache[node]
 
-                for function in shortcut_functions:
-                    if pattern_string.find(function + '(') != -1:
-                        self.shortcut_functions += [function]
+    def get_descendants(self, node):
+        return self.all_node_cache[node]
 
+class _FakeCache(object):
```

Comment:
This is just a trick. An object that emulates a cache without actually having it. As it has the same methods as the real cache, all shortcut functions can use as if there were always a cache